### PR TITLE
enhance: Add channelCPs in FlushResponse

### PR DESCRIPTION
**Pull request by wayblink**
#32609

### CODE REVIEW

**Comment by wayblink** [comment]
rerun ut

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

File: internal/datacoord/services.go [hunk 1]
```diff
@@ -81,6 +81,18 @@ func (s *Server) Flush(ctx context.Context, req *datapb.FlushRequest) (*datapb.F
 		}, nil
 	}
 
+	channelCPs := make(map[string]*msgpb.MsgPosition, 0)
+	coll, exist := s.meta.collections[req.GetCollectionID()]
```
Comment (bigsheeper) [inline internal/datacoord/services.go:85]: Please don't access the meta collections directly(without lock guard), use `s.meta.GetCollection()` instead.
By the way, you may get a nill collection here before this collection is cached in datacoord, use `handler.GetCollection()` can avoid that.

File: internal/datacoord/services.go [hunk 2]
```diff
@@ -1267,6 +1281,14 @@ func (s *Server) GetFlushState(ctx context.Context, req *datapb.GetFlushStateReq
 			if segment == nil || isFlushState(segment.GetState()) {
 				continue
 			}
+			vch := segment.SegmentInfo.GetInsertChannel()
+			cp := s.meta.GetChannelCheckpoint(vch)
```
Comment (bigsheeper) [inline internal/datacoord/services.go:1285]: group segments by vchannel to avoid get cp for each segment

File: internal/datacoord/services.go [hunk 3]
```diff
@@ -159,6 +171,7 @@ func (s *Server) Flush(ctx context.Context, req *datapb.FlushRequest) (*datapb.F
 		TimeOfSeal:      timeOfSeal.Unix(),
 		FlushSegmentIDs: flushSegmentIDs,
 		FlushTs:         ts,
+		ChannelCps:      channelCPs,
```
Comment (bigsheeper) [inline internal/datacoord/services.go:174]: Why channelCPs are needed in FlushResponse?

File: internal/datacoord/services.go [hunk 4]
```diff
@@ -1545,16 +1568,13 @@ func (s *Server) BroadcastAlteredCollection(ctx context.Context, req *datapb.Alt
...
 	// cache miss and update cache
 	if clonedColl == nil {
-		collInfo := &collectionInfo{
-			ID:             req.GetCollectionID(),
-			Schema:         req.GetSchema(),
-			Partitions:     req.GetPartitionIDs(),
-			StartPositions: req.GetStartPositions(),
-			Properties:     properties,
-			DatabaseID:     req.GetDbID(),
+		err := s.loadCollectionFromRootCoord(ctx, req.GetCollectionID())
```
Comment (bigsheeper) [inline internal/datacoord/services.go:1571]: any reason for this modification?

File: internal/proxy/task.go [hunk 5]
```diff
@@ -1439,6 +1441,15 @@ func (t *flushTask) Execute(ctx context.Context) error {
 		flushColl2Segments[collName] = &schemapb.LongArray{Data: resp.GetFlushSegmentIDs()}
 		coll2SealTimes[collName] = resp.GetTimeOfSeal()
 		coll2FlushTs[collName] = resp.GetFlushTs()
+		dcChannelCps := resp.GetChannelCps()
+		for ch, pos := range dcChannelCps {
```
Comment (bigsheeper) [inline internal/proxy/task.go:1445]: just `channelCps = resp.GetChannelCps()` ?

File: internal/datacoord/services.go [hunk 6]
```diff
@@ -1267,6 +1281,14 @@ func (s *Server) GetFlushState(ctx context.Context, req *datapb.GetFlushStateReq
 			if segment == nil || isFlushState(segment.GetState()) {
 				continue
 			}
+			vch := segment.SegmentInfo.GetInsertChannel()
+			cp := s.meta.GetChannelCheckpoint(vch)
+			channelCPs[vch] = &msgpb.MsgPosition{
```
Comment (bigsheeper) [inline internal/datacoord/services.go:1286]: just `channelCPs[vch] = cp`?

File: internal/datacoord/services.go (hunk 4 above)
Comment (wayblink) [reply internal/datacoord/services.go:1571]: req has no vchannel info, have to get it from rootcoord

File: internal/proxy/task.go (hunk 5 above)
Comment (wayblink) [reply internal/proxy/task.go:1445]: ok

File: internal/datacoord/services.go (hunk 6 above)
Comment (wayblink) [reply internal/datacoord/services.go:1286]: ok

Comment (wayblink) [reply internal/datacoord/services.go:1286]: ok

File: internal/datacoord/services.go (hunk 3 above)
Comment (wayblink) [reply internal/datacoord/services.go:174]: Backup use Flush to get FlushTs. FlushTs is used to mark the end of this backup and used to filtered data > FlushTs. CDC will use channelCPs as the beginning to transport the data in MQ between clusters. So we also support it in Flush

File: internal/datacoord/services.go (hunk 2 above)
Comment (wayblink) [reply internal/datacoord/services.go:1285]: OK

File: internal/datacoord/services.go (hunk 1 above)
Comment (wayblink) [reply internal/datacoord/services.go:85]: OK

File: internal/datacoord/services.go [hunk 7]
```diff
@@ -1267,6 +1281,14 @@ func (s *Server) GetFlushState(ctx context.Context, req *datapb.GetFlushStateReq
...
+				ChannelName: cp.GetChannelName(),
+				MsgID:       cp.GetMsgID(),
+				MsgGroup:    cp.GetMsgGroup(),
+				Timestamp:   cp.GetTimestamp(),
+			}
```
[the remark this answers is not in the document]
Comment (wayblink) [reply internal/datacoord/services.go:1291]: OK

File: internal/datacoord/services.go (hunk 4 above)
Comment (bigsheeper) [reply internal/datacoord/services.go:1574]: Will it be a better choice to modify the proto of req?

**Review by XuanYang-cn — APPROVED** [review]

**Comment by wayblink** [comment]
rerun ut

**Pull request merged**